crates/core/src/node/eth.rs

```diff
@@ -128,32 +125,34 @@ impl InMemoryNode {
                 return Err(TransparentError(err.into()).into());
             }
         } else {
-            tx_req.gas_price = tx.max_fee_per_gas.unwrap_or(U256::from(l2_gas_price));
-            tx_req.max_priority_fee_per_gas = tx.max_priority_fee_per_gas;
-            if tx_req.transaction_type.is_none() {
-                tx_req.transaction_type = Some(zksync_types::EIP_1559_TX_TYPE.into());
+            tx.gas_price = Some(tx.max_fee_per_gas.unwrap_or(U256::from(l2_gas_price)));
+            tx.max_priority_fee_per_gas = Some(tx.max_priority_fee_per_gas.unwrap_or(U256::zero()));
+            if tx.transaction_type.is_none() {
+                tx.transaction_type = Some(zksync_types::EIP_1559_TX_TYPE.into());
             }
         }
+        if tx.nonce.is_none() {
+            let nonce_key = self.storage_key_layout.get_nonce_key(&tx.from.unwrap());
+            tx.nonce = Some(h256_to_u256(self.storage.read_value_alt(&nonce_key).await?) + 1);
+        }
+
+        let mut tx_req = TransactionRequest::from(tx.clone());
+        tx_req.chain_id = Some(chain_id.as_u64());
         // Needed to calculate hash
-        tx_req.r = Some(U256::default());
-        tx_req.s = Some(U256::default());
-        tx_req.v = Some(U64::from(27));
+        tx_req.r = Some(U256::zero());
+        tx_req.s = Some(U256::zero());
+        tx_req.v = Some(U64::from(0));
```

Comment:
Given that it was a point of confusion before, maybe it makes sense to add reasoning here?